ThunderCloud/GridPage.swift

```diff
@@ -0,0 +1,105 @@
+//
+//  GridPage.swift
+//  ThunderCloud
+//
+//  Created by Simon Mitchell on 16/02/2018.
+//  Copyright © 2018 threesidedcube. All rights reserved.
+//
+
+import ThunderCollection
+import ThunderTable
+
+/// A subclass of `TSCCollectionViewController` for displaying a CMS grid page
+open class GridPage: CollectionViewController, StormObjectProtocol {
+	
+	/// An array of dictionaries which contain custom attributes fot the `StormObject`
+	public var attributes: [[AnyHashable : Any]]?
+	
+	private let dictionary: [AnyHashable : Any]
+	
+	/// The unique identifier for the storm page
+	public let pageId: String?
+	
+	/// The internal name for this page.
+	/// Named pages can be used for native overrides and for identifying pages
+	/// whose id may change with delta publishes. By default pages do not have
+	/// names but they can be added in the CMS.
+	public let pageName: String?
+	
+	public convenience init?(contentsOf url: URL) {
+		
+		guard let data = try? Data(contentsOf: url) else { return nil }
+		guard let pageObject = try? JSONSerialization.jsonObject(with: data, options: []) else { return nil }
+		guard let pageDictionary = pageObject as? [AnyHashable : Any] else { return nil }
+		
+		self.init(dictionary: pageDictionary)
+	}
+
+	required public init?(dictionary: [AnyHashable : Any]) {
+		
+		self.dictionary = dictionary
+		
+		pageName = dictionary["name"] as? String
+		
+		if let pageNumberId = dictionary["id"] as? Int {
+			pageId = "\(pageNumberId)"
+		} else {
+			pageId = dictionary["id"] as? String
+		}
+		
+		super.init(collectionViewLayout: UICollectionViewFlowLayout())
+		
+		columns = 2
+		attributes = dictionary["attributes"] as? [[AnyHashable : Any]]
+		
+		if let titleDict = dictionary["title"] as? [AnyHashable : Any], let titleContentKey = titleDict["content"] as? String {
+			title = StormLanguageController.shared.string(forKey: titleContentKey)
+		}
+	}
+	
+	required public init?(coder aDecoder: NSCoder) {
+		
+		dictionary = [:]
+		pageId = nil
```

Comment:
Rather than do this would it be better to give the properties a default value in their declaration?

Reply:
If we give the properties an initial value we get the error:
```
Immutable value 'self.dictionary' may only be initialized once
```
which is only fixable by changing the property to a `var` which I'd rather avoid for the moment :)

Reply:
Why not a var? If you make it a private var it will only be accessible inside the class.

Reply:
That's a good shout! Will do that 😆